### ISSUE TITLE
Generate properties for templates using TemplatePartAttribute

### ISSUE THREAD

**Issue by maxkatz6** [AvaloniaUI/Avalonia#11934]
There is a new attribute ported from WPF.
It will be useful to have optional code generation for "parts" of templates.
As an example, user code:
```csharp
[TemplatePart("PART_TextPresenter", typeof(TextPresenter))]
public partial class TextBox
{
    protected override void OnAttachedToVisualTree(VisualTreeAttachmentEventArgs e)
    {
        base.OnAttachedToVisualTree(e);
            
        if (IsFocused)
        {
            TextPresenter?.ShowCaret(); // TextPresenter is generated
        }
    }
}
```
Additional generated code, adds new property and sets value to it:
```csharp
#nullable enable
public partial class TextBox
{
    private TextPresenter? TextPresenter { get; set; }

    static TextBox()
    {
         TemplateAppliedEvent.AddClassHandler<TextBox>((control, args) =>
         {
             control.TextPresenter = args.NameScope.Get<TextPresenter>("PART_TextPresenter");
         });
    }
}
```

Problems:
1. Only one static ctor can be created for class, if we want to handle TemplateAppliedEvent to fill these properties.
2. Only one override of OnTemplateApplied method is possible, if we want to use it to fill these properties.

Which means, we most likely will need to modify Avalonia to add yet another way to read namescope after template was applied.

**Comment by maxkatz6** [comment]
Personally, I am leaning to this solution:
1. Generate OnApplyTemplate fully.
2. Control class is kept simple, for 90% of use cases
3. `partial void OnApplyTemplateCore` is included, which can be implemented for any extra handling _after templates_ applied. NOTE: `partial void` is important here, as partial void method don't have to have implementation (it's optional).
4. Similarly, there is a `OnTemplateApplying`, for possible cleanup of previous template parts.

I am going to add it to the next API review.

```csharp
[TemplatePart("PART_Text", typeof(TextPresenter))]
[TemplatePart("PART_AcceptButton", typeof(Button))]
public partial class TextBox
{
}
```

```csharp
public partial class TextBox
{
    private global::Avalonia.Controls.Presenters.TextPresenter Text;
    private global::Avalonia.Controls.Button AcceptButton;

    [CompilerGenerated]
    protected override void OnApplyTemplate(global::Avalonia.Controls.Primitives.TemplateAppliedEventArgs e)
    {
        OnTemplateApplying();

        Text = e.NameScope.Get<global::Avalonia.Controls.Presenters.TextPresenter>("PART_Text");
        AcceptButton = e.NameScope.Get<global::Avalonia.Controls.Button>("PART_AcceptButton");

        OnTemplateApplied(e);
    }

    partial void OnTemplateApplying();
    partial void OnTemplateApplied(global::Avalonia.Controls.Primitives.TemplateAppliedEventArgs e);
}
```

**Comment by robloo** [comment]
Right now I think @worldbeater has the best design. `InitializeTemplateParts()` is super clean and follows existing conventions.

With your [design above](https://github.com/AvaloniaUI/Avalonia/issues/11934#issuecomment-3094317635) I have the following concerns: 

 1. `OnApplyTemplate` is a very old convention that has already stood the test of time. It's really easy for control authors to do what they want within this method. It would be adding complexity to something simple and widely understood.
 2. Along the lines of point number 1: We shouldn't make devs rewrite existing code for version 12. There are a lot of controls, stuff like this should be opt-in.
 3. Your code above driving core state changes from generated code (OnTemplateApplying/OnTemplateApplied). That's a no-no in my book and this type of things should be in the main control/base class.
 4. Finally, it's adding too much complexity (OnTemplateApplying/OnTemplateApplied/OnApplyTemplateCore). I'm not against that per say but I think it's out if scope here and unnecessary to solve the problem.

My argument would be to instead take the existing approach we have with `InitializeComponent`. We will have a new `InitializeTemplateParts` method that will do all this. Devs can call this method -- or not if they choose -- from `OnApplyTemplate`. This follows the existing, very old, convention from WPF with the constructor. Everyone understands this, its simpler, and integrates side-by-side with existing code.

**Comment by maxkatz6** [comment]
@robloo I don't think my option makes devs to rewrite existing code. If they used TemplatePart attribute before, source generator doesn't have to run if user already has explicit OnApplyTemplate overriden. Plus this option follows several other source generators, primarily CommunityToolkit one.

But yeah, following InitializeComponent is a good point for consistency. Will agree there.
Personally, I don't want to add any new virtual to the Avalonia itself (like first proposed `virtual InitializeTemplateParts`). And @worldbeater option works for me too. Either way, I want to get this feature done for 12.0.

**Comment by maxkatz6** [comment]
Optionally, I would also consider adding `[TemplatePart(..., GeneratedName = "..."}]` property. That's where naming gets difficult, and extra control might be useful for some controls. `PART_ControlPresenter` will generate `ControlPresenter` or `PART_Border` will generate `Border` fields...a fuel for headache.

It will be argued whether these parts should be camel case or pascal case, "_" prefixed or not. But our name generator always creates pascal case fields, can probably ignore this problem for now, unless we want to change both generators.

**Comment by robloo** [comment]
@maxkatz6 With all your other recent work in this area is this something to be considered again?

**Comment by maxkatz6** [comment]
@robloo yes, I was actually thinking about this proposal, while writing properties source generator.
